### ISSUE TITLE
Prevent batching/buffering deadlock (`buffer.max_size` >= `batch.max_bytes`)

### ISSUE THREAD

**Issue by binarylogic** [vectordotdev/vector#355]
It is my understanding that data is not removed from disk buffers until it has been successfully sent downstream. What happens if the configured `batch.max_size` is greater than the `buffer.max_size`? The buffer will not accept any more data, but the batch will not flush until the size or time limit has been reached.

If this is true, we should prevent users from configuring disk buffers in a way that is smaller than the configured batch size.

**Comment by binarylogic** [comment]
@bruceg I'm assigning this to you to see if my concerns above are true, and gauge the amount of work needed to implement the suggested validation. If it is a lot of work I would prefer to defer this.

**Comment by bruceg** [comment]
I can confirm this is an issue, but I couldn't replicate the described behavior.

To test, I hooked up a generator source to a http sink, and fed it to a dummy HTTP server that just gave 200 responses to every request. I set up a 1,000 byte disk buffer and a 10,000 byte batch buffer with a long timeout. In that configuration, Vector sends serial requests of just over 1,000 bytes. If the disk buffer size is larger than the batch buffer, Vector sends parallel requests (up to `request.in_flight_limit`) just under 10,000 bytes long.

So, not a deadlock, but definite major performance impact.

I don't have any firm ideas on how to resolve the misconfiguration yet. It will probably require pushing down at least some of the details from the `BufferConfig` into the `SinkContext` passed to `SinkConfig::build` and then do a comparison in each sink ,presumably in a wrapper method somewhere.
